### PR TITLE
feat: add new modern.js doc support edit this page on github (#3527)

### PR DESCRIPTION
**Pull request by RyanHo97**
## Summary

Add new modern.js doc support edit this page on GitHub ([[#3527](https://github.com/web-infra-dev/modern.js/issues/3527)](https://github.com/web-infra-dev/modern.js/issues/3527))

### <samp>🤖 Generated by Copilot at c39cb48</samp>

This pull request adds an optional edit link feature to the documentation pages generated by the `doc-core` package. It introduces a new `EditLink` component and a new `_relativePath` property for the page data object. It also modifies the `EditLink` interface and the `useEditLink` hook to use a custom repository url for the edit link. It demonstrates the feature in the `doc-tools-doc` package by setting the `editLink` option in the `modern.config.ts` file.

## Details

This PR adds support for editing modern.js documentation pages on GitHub. This allows users to easily make edits to documentation and submit them as pull requests.

### <samp>🤖 Generated by Copilot at c39cb48</samp>

*  Add a new property `_relativePath` to the page data object, which is the relative path of the page file from the root directory ([link](https://github.com/web-infra-dev/modern.js/pull/3595/files?diff=unified&w=0#diff-7035015058e915e5353012c1d3fdc789c14735a85a6931d0a661c0693cf2ddeaR301), [link](https://github.com/web-infra-dev/modern.js/pull/3595/files?diff=unified&w=0#diff-3b9ef78ff832598dcdf880304579f2f3fc95e1dc287ff5a94f1d0e04b162927eR212))
*  Modify the `EditLink` interface to use a custom repository url (`docRepoBaseUrl`) instead of a pattern for the edit link ([link](https://github.com/web-infra-dev/modern.js/pull/3595/files?diff=unified&w=0#diff-b84833b1604fed5e609c7b8e25cd29620b04432b07c39fdec2bb8e6cc5a5fc2dL152-R154))
*  Create a new `EditLink` component that renders the edit link for the page, using the `useEditLink` hook to generate the final url ([link](https://github.com/web-infra-dev/modern.js/pull/3595/files?diff=unified&w=0#diff-a2b1833028d8e4e9d82c1401627067a5344b6ba9e8d18a8cd23e84cab24d5282R2), [link](https://github.com/web-infra-dev/modern.js/pull/3595/files?diff=unified&w=0#diff-a2b1833028d8e4e9d82c1401627067a5344b6ba9e8d18a8cd23e84cab24d5282R23-R30), [link](https://github.com/web-infra-dev/modern.js/pull/3595/files?diff=unified&w=0#diff-a2b1833028d8e4e9d82c1401627067a5344b6ba9e8d18a8cd23e84cab24d5282R41-R43), [link](https://github.com/web-infra-dev/modern.js/pull/3595/files?diff=unified&w=0#diff-cab956d1652a3bb6abc51e1c5449ab18ba25c565ad2fe08ae1ecd86e55b92050R1-R14), [link](https://github.com/web-infra-dev/modern.js/pull/3595/files?diff=unified&w=0#diff-94be643e9b4a39ee5fce9e0c271775a9480d471fc999baa7ffa7ab500d7cdb13R1-R37), [link](https://github.com/web-infra-dev/modern.js/pull/3595/files?diff=unified&w=0#diff-d3ee0d9d1aaf2722c77c1a5fff04da0d7ced9658eea6f78e0f480b85c50da696L2-R16))
*  Add the `EditLink` component to the `DocFooter` component, which is the footer of the documentation page ([link](https://github.com/web-infra-dev/modern.js/pull/3595/files?diff=unified&w=0#diff-a2b1833028d8e4e9d82c1401627067a5344b6ba9e8d18a8cd23e84cab24d5282R41-R43))
*  Add the `editLink` option to the `modern.config.ts` file of the `doc-tools-doc` package, which is an example documentation site using the `doc-core` package ([link](https://github.com/web-infra-dev/modern.js/pull/3595/files?diff=unified&w=0#diff-ace098861d0d64fc731554fdf597402cd4d20922a4aca5bff9697a0addb06c0dR53-R56))

## Related Issue

[#3527](https://github.com/web-infra-dev/modern.js/issues/3527)

## Checklist

- [x] I have added changeset via `pnpm run change`.
- [ ] I have updated the documentation.
- [ ] I have added tests to cover my changes.

### CODE REVIEW

File: packages/document/doc-tools-doc/modern.config.ts [hunk 1]
```diff
@@ -50,6 +50,10 @@ export default defineConfig({
           label: 'English',
         },
       ],
+      editLink: {
+        docRepoBaseUrl: 'https://github.com/test/blog/',
```
Comment (sanyuan0704) [inline packages/document/doc-tools-doc/modern.config.ts:54]: This is the test case, not the real repo url

File: package.json [hunk 2]
```diff
@@ -43,9 +43,9 @@
   },
   "engines": {
     "node": ">=14.17.6",
-    "pnpm": ">=7.12.2 <=7.29.1"
+    "pnpm": ">=7.12.2 <=8.2.0"
   },
-  "packageManager": "pnpm@7.29.1",
```
Comment (sanyuan0704) [inline package.json:48]: The pnpm version is still not match.

File: packages/cli/doc-core/src/theme-default/logic/useEditLink.ts [hunk 3]
```diff
@@ -1,14 +1,18 @@
 import { EditLink } from 'shared/types';
+import { usePageData } from '@/runtime';
+
+export function useEditLink(editLink?: EditLink) {
+  const pageData = usePageData();
```
Comment (sanyuan0704) [inline packages/cli/doc-core/src/theme-default/logic/useEditLink.ts:5]: It is a little messy there. You have already call `usePageData` and access the site config, so you don't need to add the `editLink` param.

By the way, i hope the responsibility of `useEditLink` is to provider a complete repo url and the caller doesn't need to care about some other details.So the hook would have the following optimizations:

1. delete the `editLink` param
2. join the `docRepoBaseUrl` and `relativePagePath`, take the result as the edit link.Finally return `text` and `link`.

File: packages/document/doc-tools-doc/modern.config.ts (hunk 1 above)
Comment (RyanHo97) [reply packages/document/doc-tools-doc/modern.config.ts:54]: Should I delete this paragraph or leave an empty string

Comment (sanyuan0704) [reply packages/document/doc-tools-doc/modern.config.ts:54]: You should leave a real repo url.

File: package.json (hunk 2 above)
Comment (RyanHo97) [reply package.json:48]: My mistake, I will replace the package. json version

File: packages/document/doc-tools-doc/modern.config.ts (hunk 1 above)
Comment (RyanHo97) [reply packages/document/doc-tools-doc/modern.config.ts:54]: What form or value is this real address

File: packages/cli/doc-core/src/theme-default/logic/useEditLink.ts (hunk 3 above)
Comment (RyanHo97) [reply packages/cli/doc-core/src/theme-default/logic/useEditLink.ts:5]: good，I will try to modify it this way

Comment (RyanHo97) [reply packages/cli/doc-core/src/theme-default/logic/useEditLink.ts:5]: I have made modifications
image

File: packages/cli/doc-core/src/theme-default/components/EditLink/index.tsx [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+
+export default function EditLink() {
+  const editLinkObj = useEditLink();
+
+  if (!editLinkObj) {
+    return null;
+  }
+
+  const { text } = editLinkObj;
+  const realLink = `${editLinkObj.link}`;
```
Comment (sanyuan0704) [inline packages/cli/doc-core/src/theme-default/components/EditLink/index.tsx:13]: No need to use template string.

Comment (RyanHo97) [reply packages/cli/doc-core/src/theme-default/components/EditLink/index.tsx:13]: Yes, thank you for your correction

**Review by sanyuan0704 — APPROVED** [review]

**Pull request merged**